### PR TITLE
WIP: [Experimental] Input Union

### PR DESCRIPTION
**Pull request by PascalSenn**
Adds experimental support for Input Union

### CODE REVIEW

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs [hunk 1]
```diff
@@ -144,7 +143,7 @@ public virtual NameString GetTypeName(Type type, TypeKind kind)
 
             string name = type.GetGraphQLName();
 
-            if (kind == TypeKind.InputObject
+            if ((kind == TypeKind.InputObject || kind == TypeKind.InputUnion)
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:146]: I think this is not a good idea. For input objects the name extension is a good practice but for unions? There is nothing like that established yet. Or is there?

File: src/HotChocolate/Core/src/Types/Types/InputUnionType.cs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+                                TypeResources.InputUnionType_UnableToResolveType);
+                        }
+                    }
+                }
+
+                foreach (InputObjectType type in Types.Values)
+                {
+                    if (IsInstanceOfType(type, literal))
+                    {
+                        return type.ParseLiteral(literal);
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InputUnionType.cs:100]: @michaelstaib reminder to myself ... what to do about IsInstanceOfType

File: src/HotChocolate/Core/src/Types/Types/InputUnionType.cs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+                return NullValueNode.Default;
+            }
+
+            foreach (InputObjectType type in _typeMap.Values)
+            {
+                if (type.IsInstanceOfType(value) &&
+                    type.ParseValue(value) is ObjectValueNode valueNode &&
+                    valueNode.Fields is List<ObjectFieldNode> fields)
+                {
+                    fields.Add(new ObjectFieldNode("__typename", type.Name));
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InputUnionType.cs:148]: I see what you mean ... this is really not a good solution :D

File: src/HotChocolate/Core/src/Types/Types/InputUnionType.cs [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+            try
+            {
+                if (value is null)
+                {
+                    serialized = null;
+                    return true;
+                }
+
+                //TODO: Do we have to extend with __typename?
+                if (value is IReadOnlyDictionary<string, object>
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InputUnionType.cs:178]: good question ... I guess we should.

File: src/HotChocolate/Core/src/Types/Types/InputUnionType.cs [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+                    || value is IDictionary<string, object>)
+                {
+                    serialized = value;
+                    return true;
+                }
+
+                foreach (InputObjectType type in _typeMap.Values)
+                {
+                    if (type.IsInstanceOfType(value) &&
+                            type.TrySerialize(value, out var serializedDict) &&
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Types/InputUnionType.cs:188]: indentation

File: src/HotChocolate/Core/src/Types/Utilities/NamedTypeInfoFactory.cs [hunk 6]
```diff
@@ -178,6 +178,7 @@ private static bool CanHandle(Type type)
                 || typeof(InterfaceType).IsAssignableFrom(type)
                 || typeof(EnumType).IsAssignableFrom(type)
                 || typeof(UnionType).IsAssignableFrom(type)
+                || typeof(InputUnionType).IsAssignableFrom(type)
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Utilities/NamedTypeInfoFactory.cs:181]: - [ ] we also have to update the internal class `BaseTypes`

File: src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs [hunk 7]
```diff
@@ -110,16 +123,20 @@ internal static class InputObjectParserHelper
         {
             foreach (KeyValuePair<string, object> fieldValue in source)
             {
-                if (type.Fields.TryGetField(fieldValue.Key, out InputField field))
-                {
-                    object value = field.Type.Deserialize(fieldValue.Value);
-                    target[field.Name] = ConvertValue(field, converter, value);
-                }
-                else
+                if (fieldValue.Key != "__typename")
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs:126]: we should use our introspection field constant here

Comment (michaelstaib) [reply src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs:126]: Also I think we can do that more elegantly.

File: src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs [hunk 8]
```diff
@@ -55,16 +56,28 @@ internal static class InputObjectParserHelper
             for (int i = 0; i < source.Fields.Count; i++)
             {
                 ObjectFieldNode fieldValue = source.Fields[i];
-                if (type.Fields.TryGetField(fieldValue.Name.Value, out InputField field))
+                if (IntrospectionFields.TypeName.Equals(fieldValue.Name.Value))
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs:59]: here you are using the constant but further down not.

File: src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs [hunk 9]
```diff
@@ -110,16 +123,20 @@ internal static class InputObjectParserHelper
...
+                if (fieldValue.Key != "__typename")
                 {
-                    throw new InputObjectSerializationException(
-                        $"The field `{fieldValue.Key}` does not exist on " +
-                        $"the type `{type.Name}`.");
+                    if (type.Fields.TryGetField(fieldValue.Key, out InputField field))
+                    {
+                        object value = field.Type.Deserialize(fieldValue.Value);
+                        target[field.Name] = ConvertValue(field, converter, value);
+
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Types/Utilities/Serialization/InputObjectParserHelper.cs:132]: empty lines

File: src/HotChocolate/Core/src/Validation/DocumentValidatorContext.cs [hunk 10]
```diff
@@ -107,6 +107,8 @@ public bool UnexpectedErrorsDetected
 
         public IDictionary<string, object> ContextData { get; } = new Dictionary<string, object>();
 
+        public DocumentValidatorContextPool? Pool { get; set; }
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Validation/DocumentValidatorContext.cs:110]: We should not expose the pool.

File: src/HotChocolate/Language/src/Language.Utf8/Utf8GraphQLParser.cs [hunk 11]
```diff
@@ -131,6 +131,11 @@ private IDefinitionNode ParseDefinition()
                     return ParseInterfaceTypeDefinition();
                 }
 
+                if (_reader.Value.SequenceEqual(GraphQLKeywords.InputUnion))
```
Comment (michaelstaib) [inline src/HotChocolate/Language/src/Language.Utf8/Utf8GraphQLParser.cs:134]: this should be moved down to input

File: src/HotChocolate/Language/src/Language.Utf8/Utf8GraphQLParser.TypeDefinition.cs [hunk 12]
```diff
@@ -347,6 +347,36 @@ private InterfaceTypeDefinitionNode ParseInterfaceTypeDefinition()
...
+        /// <param name="context">The parser context.</param>
+        private InputUnionTypeDefinitionNode ParseInputUnionTypeDefinition()
+        {
+            TokenInfo start = Start();
+
+            MoveNext();
+
+            NameNode name = ParseName();
+            List<DirectiveNode> directives =
+                ParseDirectives(true);
```
Comment (michaelstaib) [inline src/HotChocolate/Language/src/Language.Utf8/Utf8GraphQLParser.TypeDefinition.cs:365]: can me move this up ... this should fit in 100 chars.

File: src/HotChocolate/Core/src/Validation/IDocumentValidatorContext.cs [hunk 13]
```diff
@@ -53,6 +53,8 @@ public interface IDocumentValidatorContext : ISyntaxVisitorContext
 
         IDictionary<string, object> ContextData { get; }
 
+        DocumentValidatorContextPool? Pool { get; }
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Validation/IDocumentValidatorContext.cs:56]: This should not be exposed!

File: src/HotChocolate/Core/src/Validation/Pooling/DocumentValidatorContextPool.cs [hunk 14]
```diff
@@ -10,6 +10,13 @@ public DocumentValidatorContextPool(int size = 8)
         {
         }
 
+        public override DocumentValidatorContext Get()
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Validation/Pooling/DocumentValidatorContextPool.cs:13]: This needs to be removed!

File: src/HotChocolate/Core/src/Validation/Rules/ValueVisitor.cs [hunk 15]
```diff
@@ -207,6 +207,10 @@ public ValueVisitor()
                     }
                 }
             }
+            else if (namedType is InputUnionType inputUnionType)
```
Comment (michaelstaib) [inline src/HotChocolate/Core/src/Validation/Rules/ValueVisitor.cs:210]: The changes in this file are probably not a good idea ... we should not add so much performance implications on all the rules for a single case.

File: src/HotChocolate/Core/test/Types.Tests/HotChocolate.Types.Tests.csproj [hunk 16]
```diff
@@ -25,4 +25,8 @@
     </None>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (michaelstaib) [inline src/HotChocolate/Core/test/Types.Tests/HotChocolate.Types.Tests.csproj:28]: Remove this

File: src/HotChocolate/Core/test/Types.Tests/Types/InputUnionTypeExtensionTests.cs [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+                .AddType<FooTypeExtension>()
+                .Create();
+
+            // assert
+            FooType type = schema.GetType<FooType>("Foo");
+            Assert.Collection(type.Types.Values,
+                t => Assert.IsType<AType>(t),
+                t => Assert.IsType<BType>(t));
+        }
+
```
Comment (michaelstaib) [inline src/HotChocolate/Core/test/Types.Tests/Types/InputUnionTypeExtensionTests.cs:26]: line

File: src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs (hunk 1 above)
Comment (PascalSenn) [reply src/HotChocolate/Core/src/Types/Types/Descriptors/Conventions/DefaultNamingConventions.cs:146]: Looks like it is also like this in the spec

File: src/HotChocolate/Core/src/Core/Execution/Utilities/VariableToValueRewriter.cs [hunk 18]
```diff
@@ -59,7 +60,51 @@ internal sealed class VariableToValueRewriter
...
+                    throw new QueryException(
+                        ErrorBuilder.New()
+                            .SetMessage(TypeResources.InputUnionType_TypeNameNotSpecified)
+                            .SetCode(ErrorCodes.Utilities.TypeNotSpecified)
+                            .AddLocation(node)
+                            .Build());
+                }
+                _type.Push(inputUnion);
+                _type.Push(type);
+
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Core/Execution/Utilities/VariableToValueRewriter.cs:97]: Line

File: src/HotChocolate/Core/src/Core/Execution/Utilities/VariableToValueRewriter.cs [hunk 19]
```diff
@@ -69,13 +114,18 @@ internal sealed class VariableToValueRewriter
             INamedType namedType = _type.Peek().NamedType();
             IInputType fieldType = null;
 
+            if (IntrospectionFields.TypeName.Equals(node.Name.Value))
+            {
+                return node;
+            }
+
             if (namedType is ScalarType scalar
                 && scalar.IsInstanceOfType((ObjectValueNode)context))
```
Comment (PascalSenn) [inline src/HotChocolate/Core/src/Core/Execution/Utilities/VariableToValueRewriter.cs:123]: Line

File: src/HotChocolate/Core/test/Validation.Tests/ValuesOfCorrectTypeRuleTests.cs [hunk 20]
```diff
@@ -1162,5 +1162,138 @@ public void BadVariablesListWithInvalidItem()
                 }
             ");
         }
+
+
```
Comment (PascalSenn) [inline src/HotChocolate/Core/test/Validation.Tests/ValuesOfCorrectTypeRuleTests.cs:1166]: Line

**Comment by michaelstaib** [comment]
Let's keep this one around and see how the spec process plays out.

**Comment by michaelstaib** [comment]
Since the input union discussion moved into a completely different direction we are closing this one.

**Pull request closed without merging**